frontend/userpatch.lua

```diff
@@ -0,0 +1,154 @@
+--[[--
+Allows applying developer patches and shell scripts while running KOReader.
+
+The contents in `koreader/userpatches/` are applied on calling `userpatch.applyPatches(priority)`.
+--]]--
+
+local isAndroid, android = pcall(require, "android")
+
+local userpatch =
+    {   -- priorities for user patches,
+        early_once = "0",  -- to be started early on startup (once after an update)
+        early = "1",       -- to be started early on startup (always, but after an `early_once`)
+        late = "2",        -- to be started after UIManager is ready (always)
+                           -- 3-7 are reserved for later use
+        before_exit = "8", -- to be started a bit before exit before settings are saved (always)
+        on_exit = "9",     -- to be started right before exit (always)
+
+        applyPatches = function(priority) end, -- to be overwritten, if the device allows it.
+    }
+
+if isAndroid and android.prop.flavor == "fdroid" then
+    return userpatch
+end
+
+------------------------------------------------------------------------------------
+
+local lfs = require("libs/libkoreader-lfs")
+local logger = require("logger")
+
+-- We can not use util.lua so early during startup.
+local function removeTrailingSlash(str)
+    if str:sub(-1, -1) == "/" then
+        return str:sub(1, -2)
+    end
+    return str
+end
+-- the directory KOReader is installed in (and runs from)
+local package_dir = removeTrailingSlash(lfs.currentdir():match("^.*/"))
+-- the directory where KOReader stores user data (on SDL we may set `XDG_DOCUMENTS_DIR=~/koreader/`)
+local home_dir = removeTrailingSlash(isAndroid and android.getExternalStoragePath() or
+    os.getenv("XDG_DOCUMENTS_DIR") or package_dir)
+
+if home_dir == nil or package_dir == nil then
+    return userpatch -- live patching is not supported
+end
+
+logger.info("Live update using package_dir:", package_dir, "; home_dir:", home_dir)
+
+-- A wrapper for os.execute, can not use ffi/util early before setupkoenv.
+local function execute(...)
+    local command = {}
+    for i, v in ipairs({...}) do
+        table.insert(command, "'")
+        table.insert(command, tostring(v))
+        table.insert(command, "' ")
+    end
+    return os.execute(table.concat(command))
+end
+
+--- Run user shell scripts or lua patches
```

Comment:
what is the purpose of running arbitrary shell scripts?

Now that you have hooks available to run arbitrary lua code sounds easier to just call shell/python or perl scripts from your own lua patches.

Reply:
Executing shell scripts is a 4-liner and why not use it. If you want execute `patch ....` it would be clumsy to write a `patch.lua` and execute the `patch` command there.